### DISCUSSION TITLE
Exporting to .h does not allow for less than 8-pixels wide images

### DISCUSSION THREAD

**Discussion by IlDucci** [ImageMagick/ImageMagick#4838] [Help]
Greetings!

I've been trying to convert a pixel font into a long, 1bpp .h string. So far I've been cutting up the pieces from a grid-based texture, which was 16x16, and tried to do a second crop into the actual size, which is 6x10 pixels.

This is an 16x16 piece.
![bitmap_latinext_export_0080](https://user-images.githubusercontent.com/5641681/153753929-d3b61e77-15e8-4b8a-a532-f43922f74a2c.png)

This is the command I was using with ImageMagick-7.1.0-portable-Q8-x64 for Windows:
```RMDIR /S /Q _DUMP
MKDIR _DUMP
IM\magick.exe "bitmap_latinext.png" -crop 16x16 +repage +adjoin "_DUMP\bitmap_latinext_export_%%04d.png"
IM\magick.exe "_DUMP\bitmap_latinext_export_????.png" -crop 6x10-0-0 +repage +adjoin -flop -colors 2 -depth 1 "GRAY:_DUMP\bitmap_latinext_export_v2_%%04d.h"
COPY /b "_DUMP\*.h" EXAMPLERESULT.BIN
DEL "_DUMP\bitmap_latinext_export_????.png"
pause
```

But when I went to check if the process worked, I found out that while BMP and PNG seemed to be properly cropped into 6x10, doing the same process onto .h or headerless raw textures (.rgb, .gray) would produce 8x10 images, like this:

![Captura de pantalla 2022-02-13 134918](https://user-images.githubusercontent.com/5641681/153754055-6843d454-890c-4152-bd03-6bb4335ea63e.png)

(I used flop so the image it matched the actual direction of the glyph)

I haven't figured out yet how to put a long string of smaller files for the final result of the font because of this discrepancy. What do I have to do to in order to get 6x10 files in .h format? The source file I am trying to replace is an .h string of multiple characters with a resolution of 6x10 each.

**Comment by snibgo** [comment]
What version of IM, on what platform?

Writing small images to .h seems to work fine for me.

I suggest you paste the contents of a sample input .png and output .h here.

**Reply by IlDucci** [reply]
I just updated to the freshest version on Windows 10 x64, 7.1.0-24 Q16-HDRI, dated 2022-02-12. No changes with what I have already reported.

This is an input PNG (for the second line of my batch file that uses IM, the first one cuts a longer file into 16x16 PNG chunks and it works properly):

![bitmap_latinext_export_0080](https://user-images.githubusercontent.com/5641681/153758257-2a4071ec-821f-44b1-b66e-2fbf53a59e2a.png)

And this is what I am getting after passing that second IM line:

[bitmap_latinext_export_v2_0080.zip](https://github.com/ImageMagick/ImageMagick/files/8055742/bitmap_latinext_export_v2_0080.zip)

According to TiledGGD, it's an 8x10 image, not a 6x10.

**Comment by snibgo** [comment]
The extension ".h" usually refers to text files that can be used as C (a programming language) header files.

But you use the prefix "GRAY:" which makes an entirely different file. Using the extension ".h" for this is misleading. I strongly suggest you don't use ".h" for this purpose.

A "GRAY:" file has no header. It is just uncompressed pixels. There is no metadata describing the width or height or anything else.

Your zip file contains one file that has 10 bytes. When reading that file, we need to specify the width, height and bits per pixel. Like this:
```
magick -size 6x10 -depth 1 gray:bitmap_latinext_export_v2_0080.h info:

gray:bitmap_latinext_export_v2_0080.h GRAY 6x10 6x10+0+0 1-bit Bilevel Gray 10B 0.000u 0:00.001
```
We can successfully read the image as 6x10 pixels. We could also have read it as 10x6 pixels, or 60x1 pixels, etc.

**Comment by IlDucci** [comment]
Well, I guess that "GRAY:" thing was a happy accident on my part. I took it by reading other posts, but one of my goals was having a headerless image. I thought that by adding the .h or .rgb format, I would get a headerless image.

I need a headerless image because the program that is going to read the bitmap needs just a string of 1bpp data. When I make a long string of single ".h" files in this case, though, I get a 8 px wide headerless string, not a 6 px wide headerless string.

EDIT: To put this in another perspective:

This is the end result of that batch file, in binary form, made viewable with the TiledGGD app:
![Captura de pantalla 2022-02-13 163528](https://user-images.githubusercontent.com/5641681/153760683-fa4b79b0-d29e-4f0b-ae4f-44332386225c.png)

And this is my current endgoal, the original file I am trying to replicate:
![Captura de pantalla 2022-02-13 163617](https://user-images.githubusercontent.com/5641681/153760696-71b1f357-add3-45f9-bc75-affcea4e1ece.png)

As you can see, the difference in extra, unwanted for me width is obvious (there's also some weird fragmentation and misworking of 6px wide image as 5px that could be caused by the loss of an ending pixel or something like that, I haven't researched that part yet).

EDIT 2: So I guess my endgoal, explained in other terms, is like making a single-line tilemap for an old game, only that with this non 8x8 tile format and in binary...?

**Comment by snibgo** [comment]
> I need a headerless image because the program that is going to read the bitmap needs just a string of 1bpp data. 

What program is that? What rules does it have about rows needing to occupy an integer number of bytes? Are those rules compatible with IM's rules?

Headerless files contain no metadata. If I write a headerless image file from one program, I would not expect another program to read that file as I want.

We can use the operating system to append two copies of your ".h" file like this:
```
f:\web\im>copy /B bitmap_latinext_export_v2_0080.h+bitmap_latinext_export_v2_0080.h x.bin

bitmap_latinext_export_v2_0080.h
bitmap_latinext_export_v2_0080.h
        1 file(s) copied.
```
Then we can use IM to read x.bin as a 6x20 image:
```
magick -size 6x20 -depth 1 gray:x.bin +write info: x.png

gray:x.bin GRAY 6x20 6x20+0+0 1-bit Bilevel Gray 20B 0.000u 0:00.002
```
Here is x.png:
![x](https://user-images.githubusercontent.com/33812368/153761344-935bbd5a-5516-4079-ac95-c55896e61f3b.png)

**Reply by IlDucci** [reply]
The program is RetroArch. I'm trying to do a custom font for one of its UI (called rgui).

The original font is an actual .h file, here: https://github.com/libretro/RetroArch/blob/master/gfx/drivers_font_renderer/bitmap.h . For reference, the second picture I posted on my last message was that string of bytes converted into a binary file so I could read it with TiledGGD.

My goal is to have a similar font, but outside of the code, in .bin font. RetroArch's has additional fonts for non-Latin languages as outside files (with the width and height values placed in a code file), which they basically should follow that .h format, but binarized and externalized.



> We can use the operating system to append two copies of your ".h" file like this:
> 
> ```
> f:\web\im>copy /B bitmap_latinext_export_v2_0080.h+bitmap_latinext_export_v2_0080.h x.bin
> 
> bitmap_latinext_export_v2_0080.h
> bitmap_latinext_export_v2_0080.h
>         1 file(s) copied.
> ```

As my bat file shows, I have used almost the same command (with the exception of a wildcard to add around 500 extra pieces), and the result is shown on the first picture of my last message.

How is it possible that the actual bytes written contain an 8x10 (+ whatever amount of pieces) string of data, but ImageMagick is actually reading and detecting a 6x10 string? Is IM ignoring those two extra pixels per width?

**Comment by snibgo** [comment]
> How is it possible that the actual bytes written contain an 8x10 (+ whatever amount of pieces) string of data, but ImageMagick is actually reading and detecting a 6x10 string? Is IM ignoring those two extra pixels per width?

Yes, when IM writes the file it adds two bits of padding per row. When I read it, I tell IM it has 6 pixels per row, so it will ignore the extra two bits.

I created x.bin so I know it represents 6x20 pixels. But it has no metadata, so I need to tell IM the dimensions before it reads the file. I can lie to IM. For example, I can say it has seven pixels per row:
```
magick -size 7x20 -depth 1 gray:x.bin +write info: x.png

gray:x.bin GRAY 7x20 7x20+0+0 1-bit Bilevel Gray 20B 0.000u 0:00.002
```
IM won't guess at the dimensions of x.bin. Perhaps RetroArch is guessing, and the guess is wrong.

**Reply by IlDucci** [reply]
Then that's the problem. I don't need those two bits for what I want.

I haven't tried the font **yet** with RetroArch, I was just trying to see if I could get the same result as the original font without having to ask a programmer for a bespoke tool. Guess my luck's out with IM?

**Comment by snibgo** [comment]
I don't know what constraints you are working with. IM is (probably) the most flexible general-purpose raster image processor available. If you don't want padding for individual rows, then write just a single row. For example:
```
magick -size 6x20 -depth 1 gray:x.bin -crop 6x1 +repage +append x2.bin
```
This reads x.bin, breaks it into rows, appends those rows sideways, and writes the 120 pixels as a single row.

x2.bin is 15 bytes, which is 120 bits, so there is no padding at all.

EDIT: Files always have an integer number of bytes. So if your image size in pixels is not divisible by 8, there will be padding.
